src/transformers/trainer_utils.py

```diff
@@ -785,3 +785,32 @@ def _remove_columns(self, feature: dict) -> dict:
     def __call__(self, features: List[dict]):
         features = [self._remove_columns(feature) for feature in features]
         return self.data_collator(features)
+
+
+def check_target_module_exists(optim_target_modules, key: str):
+    """A helper method to check if the passed module's key name matches any of the target modules in the optim_target_modules.
+
+    Args:
+        optim_target_modules (`Union[str, List[str]]`):
+            A list of strings to try to match. Can be also a full string.
+        key (`str`):
+            A key to search any matches in optim_target_modules
+
+    Returns:
+        `bool` | `re.Match[str]` | `None`: True of match object if key matches any target modules from config, False or
+        None if no match found
+    """
+    if isinstance(optim_target_modules, str):
+        target_module_found = bool(re.fullmatch(optim_target_modules, key))
+    elif key in optim_target_modules:
+        # this module is specified directly in target_modules
+        target_module_found = True
+    else:
+        target_module_found = any(target_key in key for target_key in optim_target_modules)
```

Comment:
Just a small comment about readability:

We have 4 cases that are checked:
1. `optim_target_modules` is str -> regex
2. Check if `key` in `optim_target_modules` (list of str)
3. Check if any `target_key` from `optim_target_modules` is a substring of `key` (this line)
4. Regex check on each element in `optim_target_modules`

I think this could be clarified in the code a bit better by making it 4 if/else:

```python
    target_module_found = False
    if isinstance(optim_target_modules, str):
        target_module_found = bool(re.fullmatch(optim_target_modules, key))
    elif key in optim_target_modules:  # from here, target_module_found must be a list of str
        # this module is specified directly in target_modules
        target_module_found = True
    elif any(target_key in key for target_key in optim_target_modules):
        target_module_found = True
    elif target_module_found = any(bool(re.fullmatch(...
        target_module_found = True
```

Reply:
Makes sense, done!